### PR TITLE
Add GhostNetV2 (DFC attention) modules and YOLOv8 GhostV2 configs

### PR DESCRIPTION
**Pull request by kaanrkaraman**
## Summary
Add GhostNetV2 (DFC attention) building blocks and YOLOv8 GhostV2 configs, including P2/P6 variants, with stage‑specific DFC kernels.

## Changes
- Add `GhostConvV2`, `DFCAttn`, `GhostBottleneckV2`, and `C3GhostV2` modules.
- Register new modules in module exports and model parser.
- Add configs: `yolov8-ghostv2.yaml`, `yolov8-ghostv2-p2.yaml`, `yolov8-ghostv2-p6.yaml` with DFC kernels 9/7/5.

## Tests
- `python -m pytest tests/test_python.py::test_all_model_yamls`

## Notes
- DFC attention uses max‑pool downsampling and sigmoid before upsampling, per GhostNetV2 paper behavior.

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds GhostNetV2-style DFC attention building blocks and new YOLOv8 GhostV2 model YAML configs (P2 and P6 variants) for lightweight detection. 🚀

### 📊 Key Changes
-✅ Added new model configs: `yolov8-ghostv2.yaml`, `yolov8-ghostv2-p2.yaml`, and `yolov8-ghostv2-p6.yaml` (Detect heads for P3–P5, P2–P5, and P3–P6 respectively).
-🧩 Implemented new NN modules in `ultralytics/nn/modules/conv.py` and `ultralytics/nn/modules/block.py`: `GhostConvV2`, `DFCAttn`, `GhostBottleneckV2`, and `C3GhostV2`.
-🔌 Registered the new modules in `ultralytics/nn/modules/__init__.py` and enabled parsing in `ultralytics/nn/tasks.py`.
-🛠️ Adjusted YAML loading fallback in `yaml_model_load()` to pass `str(path)` in the secondary `check_yaml()` call.

### 🎯 Purpose & Impact
-⚡ Enables GhostNetV2-like DFC gating in Ghost bottlenecks, potentially improving accuracy/efficiency tradeoffs for lightweight YOLOv8-style models.
-🧱 Adds ready-to-train detection architectures (including optional P2 and P6 heads) for users targeting small-object performance (P2) or larger receptive fields (P6).
-🔄 Extends Ultralytics module registry/model parser so these layers can be used in custom YAML architectures without extra patching.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kaanrkaraman, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

This is an automated message 🤖—an engineer will assist soon. For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/nn/modules/block.py [hunk 1]
```diff
@@ -421,6 +424,102 @@ def __init__(self, c1: int, c2: int, n: int = 1, shortcut: bool = True, g: int =
...
+
+        Args:
+            c1 (int): Input channels.
+            c2 (int): Output channels.
+            k (int): Kernel size for the depthwise downsampling conv.
+            s (int): Stride.
+            attn_k (int): DFC attention kernel size.
+        """
+        super().__init__()
+        c_ = c2 // 2
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/block.py]: ⚠️ **HIGH**: `GhostBottleneckV2` sets `c_ = c2 // 2` and relies on `GhostConvV2(c_, c2, ...)` to restore `c2` channels. With the current `GhostConvV2` implementation (concat of two `c_` branches), odd `c2` values will propagate a channel mismatch through the bottleneck (and potentially into the shortcut add), causing runtime shape errors.

File: ultralytics/nn/modules/conv.py [hunk 2]
```diff
@@ -350,6 +351,39 @@ def forward(self, x):
...
+            c1 (int): Number of input channels.
+            c2 (int): Number of output channels.
+            k (int): Kernel size.
+            s (int): Stride.
+            g (int): Groups.
+            act (bool | nn.Module): Activation function.
+            dw_k (int): Depthwise kernel size for cheap operation.
+        """
+        super().__init__()
+        c_ = c2 // 2  # hidden channels
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/conv.py]: ⚠️ **HIGH**: `GhostConvV2` uses `c_ = c2 // 2` and then concatenates two `c_` tensors, so the output channel count becomes `2 * (c2 // 2)`. This silently produces the wrong number of output channels when `c2` is odd (e.g., 255 -> 254), which can break model parsing/weight shapes if someone configures odd channels.

File: ultralytics/nn/modules/block.py [hunk 3]
```diff
@@ -421,6 +424,102 @@ def __init__(self, c1: int, c2: int, n: int = 1, shortcut: bool = True, g: int =
...
+        self.bn_v = nn.BatchNorm2d(c2)
+        self.conv_h = nn.Conv2d(c2, c2, (1, k), 1, (0, k // 2), groups=c2, bias=False)
+        self.bn_h = nn.BatchNorm2d(c2)
+        self.act = nn.Sigmoid()
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Generate the DFC attention map for input features."""
+        attn = self.conv(x)
+        h, w = attn.shape[2:]
+        if h > 1 and w > 1:
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/block.py]: 💡 **MEDIUM**: `DFCAttn.forward()` gates pooling on `h > 1 and w > 1`, but the actual pooling kernel is `pool_stride` (default 2). If `pool_stride` is configured > 2 (or > min(h, w)), `MaxPool2d(kernel_size=pool_stride, ...)` will throw for small feature maps. Consider gating on `h >= pool_stride and w >= pool_stride` (or dynamically choosing a safe kernel) to avoid runtime errors for small inputs / later stages.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Solid integration wiring (module registration + parser support) and the YAML heads/backbones look internally consistent. Main concern is a correctness bug in `GhostConvV2`/`GhostBottleneckV2` for odd `c2` values (output channels become `2*(c2//2)`), which can lead to silent shape mismatches or runtime add/concat failures. `DFCAttn` pooling guard is also slightly unsafe if `pool_stride` is ever set larger than the feature map.

💬 Posted 3 inline comments

File: ultralytics/nn/modules/block.py (hunk 1 above)
Comment (kaanrkaraman) [reply ultralytics/nn/modules/block.py]: [59aaf5e](https://github.com/ultralytics/ultralytics/pull/23108/commits/59aaf5e0005582abcca0772cfff758ca51f7c759)

File: ultralytics/nn/modules/conv.py (hunk 2 above)
Comment (kaanrkaraman) [reply ultralytics/nn/modules/conv.py]: [59aaf5e](https://github.com/ultralytics/ultralytics/pull/23108/commits/59aaf5e0005582abcca0772cfff758ca51f7c759)

**Comment by kaanrkaraman** [comment]
Can someone review this?

**Comment by glenn-jocher** [comment]
Thanks for the follow-up—this PR is in the review queue and someone from the team will take a look when available (no ETA); in the meantime, please make sure all GitHub Actions CI checks are green and consider adding a short benchmark/accuracy note vs a baseline to help reviewers evaluate the change per our [Contributing guide](https://docs.ultralytics.com/help/contributing/).

**Comment by glenn-jocher** [comment]
Thanks for the careful work implementing GhostNetV2's DFC attention and wiring up the GhostConvV2/GhostBottleneckV2/C3GhostV2 modules and configs — the code is clean and well documented. We're going to close this one, though: it adds a substantial new attention/architecture stack (300+ lines and three new YAMLs) to the legacy YOLOv8 family, and per our "less is more" philosophy we hold new architecture modules to a high bar of demonstrated benefit. We asked earlier for a benchmark/accuracy comparison against a baseline to justify the addition, and without that evidence we can't evaluate the accuracy/efficiency tradeoff. If you're able to gather training results (mAP + speed/params vs a matched YOLO11/YOLO26 baseline) showing a clear win, we'd be glad to revisit — please don't let this discourage you from contributing again. Thanks again — please don't hesitate to open a new PR if you have benchmark-backed results to share! 🚀

**Comment by glenn-jocher** [comment]
Thanks for the effort here, @kaanrkaraman! We're not landing the GhostNetV2 modules themselves yet — new backbone blocks need benchmark evidence (mAP/latency vs the current YOLO backbones on a real dataset) before we add them to the library. But we've reduced this branch to a zero-diff so we can still merge it and get you into the contributor list. 🙌 If you can share GhostNetV2 results that show a favorable accuracy/latency trade-off, we'd genuinely welcome a fresh PR — thanks for building with Ultralytics!
